### ISSUE TITLE
Array class is generic, but not safe

### ISSUE THREAD

**Issue by MikOfClassX** [libgdx/libgdx#7026]
Array class is generic, but not safe.
It can be used anywhere from a libgdx app but it may be source of very dangerous issues i.e. using sort() methods even from different instances from different threads of the Array class (say, the EDT and the GL loops at the same time).

Here's the fix.

https://github.com/libgdx/libgdx/commit/31102f4a0261b626f3155ed06fe4e2741b6c9d25

**Comment by tommyettinger** [comment]
I can kinda see this being an issue, but to me it doesn't make any sense to keep a Sort instance in every Array. It makes more sense to have a Sort instance in every thread that needs sorting, and then you can just use that instance to sort with [Sort.sort(Array)](https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/utils/Sort.java#L30-L32). The other option would be a stateless (fully-static) sorting algorithm like at least one used by the fastutil library, though I don't know how that compares to TimSort on performance.

**Comment by MikOfClassX** [comment]
@tommyettinger I checked the impact of having a Sort instance in every Array 

`private final Sort sorter = new Sort();`
Looking at the Sort class, data per instance is limited to 2 empty object fields (class references, not initialized at ctor() time). I guess it's no more than few bytes. References have a typical size of 4 bytes on 32-bit platforms and on 64-bits platforms with heap boundary less than 32Gb (-Xmx32G), and 8 bytes for this boundary above 32Gb.

In my opinion the extra payload of the Array with per-instance Sort class is worth thread safety (3 extra obj references, or 24bytes). No need to use per-thread storage. Stateless may be useful.

Anyway I saw that lazy instantiation rules in the Array and Sort class. So this can be easily implemented in order to remove the per-instance Sort class. Kind-of overkill to me, but ok.

btw I have my good old stateless thread-safe fast quicksort. It may be of your interest.

**Comment by tommyettinger** [comment]
Honestly I don't know the reason why we have a copy of Android's TimSort class in libGDX... We do need to make sure the sort is stable if TimSort is stable, but as long as we ensure that, I think we could swap out the Sort implementation for a stateless one without breaking anything. Then we wouldn't need any references anywhere, and threads could sort however they like, woo! I'd love to set up some benchmarks for various sorting algorithms; there are lots of factors to consider.

**Comment by MikOfClassX** [comment]
So you're inviting me to submit you my code eh ? :)

**Comment by tommyettinger** [comment]
If you want! I think I can find at least a few sorts on my own. The one I'm curious about is adapted from fastutil:

> Sorts the specified range of elements according to the order induced by the specified comparator using mergesort.
This sort is guaranteed to be stable: equal elements will not be reordered as a result of the sort. The sorting algorithm is an in-place mergesort that is significantly slower than a standard mergesort, as its running time is O(n (log n)^2), but it does not allocate additional memory; as a result, it can be used as a generic sorting algorithm.

There's of course the JDK `Arrays.sort()` methods, which can often allocate or switch to a parallel implementation behind-the-scenes. Checking Wikipedia, Blocksort looks just about optimal, but has lots of hidden constant factors in its Big-O analysis.

Also checking Wikipedia, it looks like Timsort is stable, but Quicksort rarely is, so changing to most Quicksort implementations would be a breaking change (and we'd want to keep Timsort for cases where stability is mandatory).

**Comment by MikOfClassX** [comment]
So you already have a verdict, and I mostly agree with you.
I can do the lazy per-instance Sort() into arrays and this wil be a very good tradeoff.
What do you think ?

**Comment by tommyettinger** [comment]
I'm just not sure if sorting methods or objects even belong in Array, since you can already call methods to sort an Array on a Sort object. Users can have full control over Sort objects (regarding threads), but we couldn't remove the local Sort instances from Array code once they're in the API (and subclasses would always inherit them)... I'm going to want to hear from more of the team; threading is not a strength of mine.

**Comment by MikOfClassX** [comment]
Yep, static sorting class into Array seems kind-of extra. The Sort class comment itself says it all : 

/** Returns a Sort instance for convenience. Multiple threads must not use this instance at the same time. */
..So we know we must not use it eh.

Afaics libgdx classes are generally missing encapsulation and this is makig almost impossible for the maintainers to mod the inner workings without breaking someone's code.

**Comment by saphirelsnow** [comment]
I think Encapsulation was a goal for Android 13 compatibility?

**Comment by tommyettinger** [comment]
OK, this is... interesting... TimSort is used libGDX in its `Sort` class, but is also used by `Arrays.sort()` in the JDK when sorting Comparable types or when given a Comparator for Object items. I don't see any reason so far why we have a copy of TimSort from Android's JDK here; we can just use `Arrays.sort()` and it has the same properties (it may allocate a little extra per-sort, but nothing compared to the up-to-`O(n)` that TimSort can allocate as working space). We should probably be using `Arrays` if it is present on all platforms now, and we can call that statically from our `Array` class unless allocating temporary TimSort instances is too expensive.

[I did run some benchmarks](https://github.com/tommyettinger/assorted-benchmarks/blob/master/squid/src/main/java/com/github/tommyettinger/squidlib/SortBenchmark.java#L182-L202); JDKSort and GDXSort are always very close because their code is nearly identical. EttingerSort works on a List<String> instead of a String[], so it is hard to compare, but DSSort is an in-place merge sort that doesn't allocate working space and operates on an array; it has `O(n log(n) log(n))` worst-case runtime where TimSort has `O(n log(n))`, and the benchmark tests a worst-case. DSSort is the same as FastUtilMergeSort, or at least is extremely similar (it doesn't use the same IntComparator and Swapper combination that FastUtil does). FastUtilQuickSort is, strangely, extremely slow here, both in serial and parallel versions.

**Comment by NathanSweet** [comment]
I couldn't remember why we have our own TimSort, but some searching dug up chat conversation from 2010. It all started with: "ryanm claims Arrays.sort allocates", so it's definitely ryanm's fault! Our fork of TimSort keeps its allocations so subsequent sorts may avoid allocation. Is that worthwhile and/or a good idea these days? Hard to say. Probably?

FWIW, sorting can be difficult to compare, depending a lot on what is being sorted (random, nearly sorted, sorted backward, etc). TimSort is supposed (and I guess is proven) to work well in real world scenarios a lot of the time. I'm open to changing it, as we all know @tommyettinger can certainly put together a compelling argument, but I expect it to be difficult to prove a replacement is better in general than TimSort.

Regarding threading, in retrospect Array shouldn't have sort methods and Sort shouldn't have a static instance. At this point it's probably best to leave it, pretend that is the case, and manage your own Sort instances.

Regarding encapsulation, library maintainers love it, users trying to get things done hate it. We've ended up somewhere in the middle, though it varies in different parts if libgdx.

**Comment by tommyettinger** [comment]
Arrays.sort() does allocate, but so does TimSort -- that's because Arrays.sort() uses TimSort, and has since Java 7. I can look more closely at the history of TimSort in libGDX, but it looks like a pretty clear copy of code that's in the JDK (and even though Android has it, it probably is legally considered GPL v2, like OpenJDK). It might be Sort.java where the code diverges; Arrays in the JDK seems to have a thin wrapper around code that creates a TimSort or ComparableTimSort object, and that's similar to Sort in libGDX. The in-place merge sort I benchmarked shouldn't allocate more than O(1) space, but TimSort needs working space because it isn't in-place, and can use O(n) space in the worst case (specifically, if the data being sorted is randomly ordered and has 1000 objects, TimSort can allocate up to 500 objects in an array to use as temporary storage).

I think maybe the best solution would be to retain any public classes that are part of the API, but change Sort's static methods to use java.util.Arrays internally, just for thread sanity purposes. That would make the sort() in Array that uses the Sort class thread-safe, without changing any members in Array, without changing any method signatures, and even without changing the algorithm used for sorting. Unfortunately, this does allocate a TimSort behind the scenes, but that has to be done anyway to resize the working data and make sure it is the right type, as far as I can tell. There's also the possibility that GWT or other platforms could implement a JDK class more efficiently for their specific platform, since Arrays.sort() only requires a stable sort, not a particular algorithm.

**Comment by tommyettinger** [comment]
Ugh, this code is complex. Okay, (Comparable)TimSort does have an optimization where it can resize or reuse a backing array for temporary working space. That's good unless you need multiple threads, in which case they would fight over the same working space if a Sort were static... I think most people working with threads in libGDX are already on shaky ground, and should be reading docs and source code heavily to ensure their code will work across threads. Keeping that in mind, it might make the most sense to change static usage of Sort to use Arrays instead, but leave Sort otherwise unchanged so users who need threads can have one Sort per thread and so avoid any contention.

**Comment by MikOfClassX** [comment]
I'd tend to stick to @tommyettinger idea of a per-thread sort, which is both simple, fast and inexpensive.
Our per-instance impl is already ok, also considering that an extra per-instance payload is 24bytes if you don't sort() and some more if you use the sort().
A thread-local will give us thread safety and efficiency at almost-zero impact and no api change.

About encapsulation, I'm not sure users will hate it. I like to think that libgdx adopters are used to oop (they use java in the end). And maintainers will be definitely free to rename a field without breaching anyone's apps.
Now, I think that refactoring the whole lib is unadvisable. I'm under the impression that libgdx will remain as it is now.
